### PR TITLE
[nn] TransformerEncoderLayer: no-batch-dim

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: https://github.com/pytorch/pytorch/issues/60585
TODO:
* [ ] Update docs?
* [x] Generic reference function?

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### CODE REVIEW

**Comment by kshitij12345** [comment]
[Docs](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html) redirect the user to look at `Transformer` documentation to check for shape.

Should copy the shape to `TransformerEncoderLayer` or update `Transformer` documentation once we update it?

**Comment by jbschlosser** [comment]
> [Docs](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html) redirect the user to look at `Transformer` documentation to check for shape.
> 
> Should we copy the shape to `TransformerEncoderLayer` or update `Transformer` documentation once we update it (we will update `TransformerDecoderLayer` followed by `Transformer` hopefully in a week or two)?

Imo filling out the shape for `TransformerEncoderLayer` can be done in a separate doc PR

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -232,18 +232,42 @@ def no_batch_dim_reference_fn(m, p, *args, **kwargs):
...
-    reduction = get_reduction(m)
-    if reduction == 'none':
-        return output.squeeze(0)
-    # reduction is 'sum' or 'mean' which results in a 0D tensor
+        output = m(*single_batch_input_args, **kwargs).squeeze(batch_dim)
+
+    if is_criterion:
+        reduction = get_reduction(m)
+        if reduction == 'none':
+            return output.squeeze(0)
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:270]: This looks pretty good!

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -440,8 +492,7 @@ def module_inputs_torch_nn_MultiheadAttention(module_info, device, dtype, requir
                module_inputs_func=module_inputs_torch_nn_Hardswish,
                supports_gradgrad=False),
     ModuleInfo(torch.nn.TransformerEncoderLayer,
-               module_inputs_func=module_inputs_torch_nn_TransformerEncoderLayer,
-               supports_gradgrad=False),
+               module_inputs_func=module_inputs_torch_nn_TransformerEncoderLayer),
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:495]: Was `supports_gradgrad=False` intentionally removed?

**Review by jbschlosser — COMMENTED** [review]
Awesome! Just one thing I'm not sure about but otherwise looks great :)

File: torch/testing/_internal/common_modules.py (hunk 2 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_modules.py:495]: Yeah. I didn't see any comment mentioning why it was added in https://github.com/pytorch/pytorch/pull/64444 and the test worked so.

**Review by jbschlosser — APPROVED** [review]
LGTM thanks!
